### PR TITLE
Gui: improve NaviCube rendering (simplified)

### PR DESCRIPTION
**Pull request by realthunder**
Related #6689 #6839

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6848">pipeline status</a> for feature branch [PR_6848](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6848). Pipeline [533777842](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/533777842) was triggered at [a06d76a](https://github.com/FreeCAD/FreeCAD/pull/6848/commits/a06d76a4a2d67893ab07ce8368c3293ccfafad98). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by donovaly** [comment]
Many thanks! I will merge on Wednesday when I am back home.

I will also check if @Jolbas suggestions from #6839 can be applied too.

File: src/Gui/NaviCube.cpp [hunk 1]
```diff
@@ -269,6 +275,8 @@ NaviCubeImplementation::NaviCubeImplementation(
 	OnChange(*hGrp, "HiliteColor");
 	OnChange(*hGrp, "ButtonColor");
 	OnChange(*hGrp, "CubeSize");
+	OnChange(*hGrp, "BorderWidth");
+	OnChange(*hGrp, "BorderColor");
```
Comment (donovaly) [inline src/Gui/NaviCube.cpp:279]: Why is the BorderColor important to fix the rendering? The PR fixes the issue also without a border color.

I committed this now nevertheless but please let us only change things that have to be changed to fix an issue.

File: src/Gui/NaviCube.cpp [hunk 2]
```diff
@@ -301,7 +309,11 @@ void NaviCubeImplementation::OnChange(ParameterGrp::SubjectType &rCaller, Parame
 		m_ButtonColor.setRgba(rGrp.GetUnsigned(reason, QColor(226, 233, 239, 128).rgba()));
 	} else if (strcmp(reason,"CubeSize") == 0) {
 		m_CubeWidgetSize = (rGrp.GetInt(reason, 132));
-	}
+	} else if (strcmp(reason,"BorderWidth") == 0) {
+		m_BorderWidth = rGrp.GetFloat("BorderWidth", 1.5);
```
Comment (donovaly) [inline src/Gui/NaviCube.cpp:313]: This looked clumsy for me, so I had to change it from 1.5 to 1.1

**Comment by donovaly** [comment]
I committed this now plus some minor fixes and some of the suggestions by @jolbas: commit f70329390cea.

However, we now have a minor regression:
- set the cube size in the preferences to 500
- result:
FreeCAD_TDV64QZpsd

So the corner frame is thicker than the square lines. It should have the same width. Can you please have a look? I could not yet find a solution.

**Comment by donovaly** [comment]
> I committed this now plus some

-> closing this PR as being merged

**Comment by donovaly** [comment]
> I committed this now plus some minor fixes and some of the suggestions by @Jolbas: commit [f703293](https://github.com/FreeCAD/FreeCAD/commit/f70329390ceaac95200d6a58b1a6ff2c4e9f3ed5).
> 
> However, we now have a minor regression:
> 
>     * set the cube size in the preferences to 500
> 
>     * result:
>       FreeCAD_TDV64QZpsd
> 
> 
> So the corner frame is thicker than the square lines. It should have the same width. Can you please have a look? I could not yet find a solution.

It is really weird because when I restart FC, I get every time a different width for the corner frame. Sometimes it is correct for a 500 px cube, but not for a 150 px cube, sometimes it is the opposite. Therefore I think we cannot do much.

**Comment by realthunder** [comment]
The corner border is still drawn with texture. When the cube size is changed, the texture is not recreated, but just scaled to match the new cube size. That's why you see the corner border got thickened. When you open a new document, the texture size is determined by the new cube size, so no scaling there. This is one of the problem fixed in my first NaviCube PR.

In theory, we can use opengl to draw the triangle borders using various cube face vertices. I didn't do that, because I thought the triangle corner face is too small to click easily, especially if it is not front facing. So I reverted it back to the circular face as in 0.19. Please note that I did all these a while ago, so is the border color addition. I just wanted to minimize changes (against my existing implementation) to avoid unexpected error. This is the reason I included the corner shape change in my second NaviCube PR.

Allow me to make some complain here. Lack of support of anti-alias is arguably not a bug in the first place. I don't understand why you have to be so strict about bug vs. feature distinguish, while in reality many changes don't have clear cut differentiation. I doubt any one will complain about unfairness because of those changes. I thought this feature close thingy is kind of flexible. You don't see me complain when some of my feature PRs didn't make it despite being submitted in time.

File: src/Gui/NaviCube.cpp [hunk 3]
```diff
@@ -380,8 +393,6 @@ GLuint NaviCubeImplementation::createCubeFaceTex(QtGLWidget* gl, float gap, cons
 		QPainterPath pathSquare;
 		pathSquare.addRect(QRectF(gapi, gapi, (qreal)texSize - 2.0 * gapi, (qreal)texSize - 2.0 * gapi));
 		paint.fillPath(pathSquare, Qt::white);
-		paint.setPen(pen);
-		paint.drawPath(pathSquare);
 	}
 	else if (shape == SHAPE_CORNER) {
 		QPainterPath pathCorner;
```
Comment (Jolbas) [inline src/Gui/NaviCube.cpp:398]: Couldn't comment on line 399 but try change line 399 to
```
                qreal midTex = (qreal)texSize / 2.0;
		QRectF rectCorner = QRectF(midTex - sqrt(0.5) * gapi, midTex - sqrt(2/3) * gapi, sqrt(2) * gapi, sqrt(1.5) * gapi);
```
or possibly
```
                qreal midTex = (qreal)texSize / 2.0;
		QRectF rectCorner = QRectF(midTex - sqrt(0.5) * gapi, midTex - sqrt(1/6) * gapi, sqrt(2) * gapi, sqrt(1.5) * gapi);
```
Some of these two should be right, but I can't test by myself.

Comment (donovaly) [reply src/Gui/NaviCube.cpp:398]: The first leads to this:
FreeCAD_SB1ac5f1fy

the second one to this:
FreeCAD_nsJOFeAvt7

**Comment by donovaly** [comment]
> When you open a new document, the texture size is determined by the new cube size, so no scaling there

Thanks, I understand. But since the cube size is rarely changed, we can live with this I think.

> because I thought the triangle corner face is too small to click easily, especially if it is not front facing. So I reverted it back to the circular face as in 0.19. 

Allow me to complain here ;-): There was a decision for 0.20 to use corner triangles because this way one cannot accidentally click the faces from behind. This was a problem of the old cube. The change was discussed and now 0.20 is almost done. Therefore I don't want to have a change that reverts a decision. My approach is to discuss these changes with the community. What I learned the hard way during the 0.19 release process is that there are many displays around and since different users have different needs, decisions that seem obvious, can quickly introduce issues for other users.
Therefore every design change should be discussed. (We have users of RTL languages, different resolutions, color blind persons, trackball users... )

> Lack of support of anti-alias is arguably not a bug in the first place.

This issue popped up since 0.18 when I remember correctly. Also during the new cube development is popped up and now again. And I can confirm the look was ugly on small Hi-DPI laptop screens. So it is an issue that affect different people and it is something you immediately notice when starting FC the first time, there fore also a UX issue.

> I doubt any one will complain about unfairness because of those changes. 

I hope so. My background is 10 years of the project lyx.org and there people often went crazy when even small features were introduces in freeze times. Now it is the first time I am a release manager at all, so I have to learn.
I am happy that you are not upset that not all your PRs went in despite sent in time.

File: src/Gui/NaviCube.cpp (hunk 3 above)
Comment (Jolbas) [reply src/Gui/NaviCube.cpp:398]: Then we have to leave it. I did test the two versions out of scope along with the original. And my first version returns very similar results to the original as long as the gap is set to 0.12. Clearly there is some side effects I don't understand. Thank you. I need to set up my own build system before I go further.

**Comment by adrianinsaval** [comment]
This is kinda not the place to discuss it but I think I agree with realthunder here, perhaps there is space to be a little more flexible for cases like this were the line is blurred and it's hard to separate fix from feature, I totally understand uwe's reservations considering how sensitive some people can get though. (And i'll take this opportunity to once again thanks both of you for all the hard work you put into this project)

Regarding the corner shape, IIRC the problem was that the chamfered edges didn't have filling and weren't clickable, and once they were filled you preferred triangles over circles because they fit the shape better, I could easily be wrong here but I don't remember any technical reason why triangles were preferred over circles, personally I like this recent request to make the vertex hexagonal instead triangular: https://forum.freecadweb.org/viewtopic.php?f=34&t=68544 they are more functional than triangles and IMO look better than circles and triangles, further discussion is probably better done either there or on a new thread so the non dev part of the community also engages.

**Comment by donovaly** [comment]
> I don't remember any technical reason why triangles were preferred over circles

Geometry - the problem was that the circles can accidentally clicked from behind since they will always protrude over the actual cube. I had a look what other CAD programs do and this is for example how it is done in OCC's CAD Assistant:
firefox_EqPYiDex4S

This is also a proposal that should do the job:
https://forum.freecadweb.org/viewtopic.php?p=593838#p593838

However, 0.20 has to be done and this is feature-froozen.

**Pull request closed without merging**